planner/core/logical_plans.go

```diff
@@ -313,6 +386,123 @@ type LogicalProjection struct {
 	AvoidColumnEvaluator bool
 }
 
+// ExtractFD implements the logical plan interface, extracting the FD from bottom up.
+func (p *LogicalProjection) ExtractFD() *fd.FDSet {
+	// basically extract the children's fdSet.
+	fds := p.logicalSchemaProducer.ExtractFD()
+	// collect the output columns' unique ID.
+	outputColsUniqueIDs := fd.NewFastIntSet()
+	notnullColsUniqueIDs := fd.NewFastIntSet()
+	outputColsUniqueIDsArray := make([]int, 0, len(p.Schema().Columns))
+	// here schema extended columns may contain expr, const and column allocated with uniqueID.
+	for _, one := range p.Schema().Columns {
+		outputColsUniqueIDs.Insert(int(one.UniqueID))
+		outputColsUniqueIDsArray = append(outputColsUniqueIDsArray, int(one.UniqueID))
+	}
+	// map the expr hashCode with its unique ID.
+	for idx, expr := range p.Exprs {
+		switch x := expr.(type) {
+		case *expression.Column:
+			continue
+		case *expression.CorrelatedColumn:
+			// t1(a,b,c), t2(m,n)
+			// select a, (select c from t2 where m=b) from t1;
+			// take c as constant column here.
+			continue
+		case *expression.Constant:
+			hashCode := string(x.HashCode(p.ctx.GetSessionVars().StmtCtx))
+			var (
+				ok               bool
+				constantUniqueID int
+			)
+			if constantUniqueID, ok = fds.IsHashCodeRegistered(hashCode); !ok {
```

Comment:
What would happen for the `ok` case?
If the same constant appears in different operators but corresponds to different `UniqueID`, will there be mistakes?

Reply:
The same constant will use the same hashcode, then use the same unique id here.

This is temp solution to fix it before we refactor the way we store the const.